### DISCUSSION TITLE
Some questions on CURLOPT_ALTSVC

### DISCUSSION THREAD

**Discussion by boytsovea** [curl/curl#7954] [Q&A] [answered]
Hi guys.

We are experimenting with libcurl's http/3 support in combination with multi-interface and have some questions on how it is supposed to operate. Our libcurl binary is built with both http/2 and http/3 support enabled. 

1. We want to know, whether using the same altsvc-cache file from multiple threads or multiple processes is safe (synchronized by libcurl itself in some way) or not (and thus some external synchronization is required). If it is not thread/process-safe, is there some point, at which our code can reliably provide such synchronization?

2. In our experiments we observed that switching (or not switching) to http/3 via altsvc with multi-interface depends on lifetime of the multi-object. Consider the following example:
~~~c
#include <curl/curl.h>
#include <iostream>

int main(int, char**)
{
   CURLM* multi_handle = curl_multi_init(); // long-living multi handle
   for( int i = 0 ; i < 5 ; ++i )
   {
     //CURLM* multi_handle = curl_multi_init(); // short-living multi handle
     CURL* h;
     int still_running;

     h = curl_easy_init();

     curl_easy_setopt(h, CURLOPT_URL, "https://nghttp2.org/");
     curl_easy_setopt(h, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_NONE);
     curl_easy_setopt(h, CURLOPT_ALTSVC, "altsvc-cache.txt");
     curl_easy_setopt(h, CURLOPT_VERBOSE, 1l);
     
     curl_multi_add_handle(multi_handle, h);
     curl_multi_perform(multi_handle, &still_running);

     do
     {
       int numfds = 0;
       CURLMcode mcode = curl_multi_poll(multi_handle, NULL, 0, 1000, &numfds);

       if(mcode)
       {
         std::cerr << "curl_multi_wait() failed, code " << mcode << std::endl;
         break;
       }
       curl_multi_perform(multi_handle, &still_running);

     }
     while(still_running);

     curl_multi_remove_handle(multi_handle, h);
     curl_easy_cleanup(h);
     //curl_multi_cleanup(multi_handle);
   }
   curl_multi_cleanup(multi_handle);
   return 0;
}
~~~
If we use one long-living CURLM object for all five request with initially empty altsvc-cache.txt, than all requests are delivered with http/2. But if we use short-living CURLM objects, than the first request is delivered with http/2 and subsequent 4 requests are delivered with http/3. In our app we use long-living CURLM objects, but desired behaviour is switching to http/3 as fast as possible. Can we achieve that without recreating of CURLM objects? Are there any ways to force CURLM object to reload altsvc cache from external storage?

**Comment by bagder** [comment, marked as answer]
> We want to know, whether using the same altsvc-cache file from multiple threads or multiple processes is safe

Since you should not use the same handle in more than one thread concurrently according to the libcurl thread-safe rules, there's no such way provided. One way to provide that ability could be to add support for the alt-svc cache to the share object.

> switching (or not switching) to http/3 via altsvc with multi-interface depends on lifetime of the multi-object

Why?

**Reply by boytsovea** [reply]
> Since you should not use the same handle

We are not using the same handle from multiple threads. The question is whether we can use the same value for CURLOPT_ALTSVC for different handles in different threads and processes (or not)? I suppose that we can't, but that means, that we should somehow bind specific cache file to specific handle (and thus implement some logic to distribute some "pool" of such files to CURLM objects) or elect some "master" CURLM object, which will populate the cache (but in combination with long-living CURLM objects in other threads/processes this approach will require some mechanism to reload the contents of cache file on demand).


> Why?

We observe this behaviour in the provided example. It is unexpected (in our opinion).

**Reply by bagder** [reply]
> The question is whether we can use the same value for CURLOPT_ALTSVC for different handles in different threads and processes

Right that would be tricky, as they would all write their caches to that file independently without knowing about the others.

You would perhaps rather prefer callbacks for reading/writing the cache, much like we have for HSTS...

> We observe this behavior in the provided example. It is unexpected (in our opinion).

Then maybe we should start there? It sounds like a bug to me, so with more info and details around this problem we might understand this better!

**Reply by boytsovea** [reply]
> Then maybe we should start there? It sounds like a bug to me

Look at the code in my initial post. There are two CURLM objects:

- "long-living" one which is created at line 6 (just before the loop) and destroyed at line 42 (after the loop). Then this CURLM object is in use, we observe in cerr output (I can provide full output, if it is required), that all five transfers inside the loop use HTTP/2
- "short-living" one which is created at line 9 and destroyed at line 40 (it is commented out in provided code sample). If we comment out "long-living" CURLM object and uncomment "short-living" one, than we observe, that the first request is served using HTTP/2, than it populates altsvc cache and than subsequent 4 requests are delivered using HTTP/3. 
In my understanding of libcurl's api design the second use case (short-living CURLM) is rather strange.

Whether the observed behavior in the first use case is a bug or not is not immediately clear to me. From the one hand, since the first transfer had already populated altsvc cache, it is "logical" to use more "cool and modern" HTTP/3 for subsequent transfers, than "not so cool and not so modern" HTTP/2. From the other hand, after the first transfer is complete, "long living" CURLM object has established and ready for reuse HTTP/2 connection with all possible handshakes complete and some associated state. It also may seem "logical" to reuse it than try to establish new "logical" connection with HTTP/3. So here I'm asking you for clarification: what is the expected behavior from your point of view?

**Reply by bagder** [reply]
I think this would be a much better discussion if we actually talked about exactly what happens over the protocols. I presume this:

You fire off five transfers at once, and they all create one connection each (due to no pipewait) so you get five HTTP/2-using connections getting alt-svc data back. But by then all five transfers are already setup and using HTTP/2. When they are done, the altsvc cache is saved to disk so that the next time this function is called, it can load the cache and then use HTTP/3 for all five transfers.

**Reply by boytsovea** [reply]
> You fire off five transfers at once

The program in the initial post issues all requests sequentially: the next one is initiated after the previous one is complete. If I run this program twice, than the second run issues all request with HTTP/3, since it finds corresponding records in altsvc-cache.txt, produced by the previous run. I've attached output of the first run (output.txt), where you can see, that all five request are sequential and all use the same HTTP/2 connection, which was established by the first request (I've added code `std::cerr << "request # " << i << " is complete" << std::endl;` at line 40 of the initial example to explicitly show loop iterations).
[output.txt](https://github.com/curl/curl/files/7473024/output.txt)

**Comment by bagder** [comment]
The key is likely in this snippet:
~~~
* Alt-svc connecting from [h2]nghttp2.org:443 to [h3-29]nghttp2.org:443
* Found bundle for host nghttp2.org: 0x1166558 [can multiplex]
* Re-using existing connection! (#0) with host nghttp2.org
~~~

It finds an alt-svc match but still decides to reuse an already existing connection rather than creating a new one. Reuse is usually the fastest way to get the transfer started.

We should *probably* consider an option or something that says something like "prefer strict alt-svc before reuse".

**Comment by boytsovea** [comment]
> Reuse is usually the fastest way to get the transfer started.

Ok, thanks for clarification. One more question. If  I change the code at line 16 from 
```C++
curl_easy_setopt(h, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_NONE);
```
to
```C++
if( !i )
   curl_easy_setopt(h, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_NONE);
else
   curl_easy_setopt(h, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_3);
```

that is for all loop iterations, except the first one, try to explicitly ask for usage of HTTP/3, I see very strange output. The first request is delivered using HTTP/2 as it is expected to be. But subsequent ones are delivered with headers indicating HTTP/3, but via existing HTTP/2 (see output.txt) connection (and inspecting network traffic with Wireshark confirms this observation). It looks like a bug.
[output.txt](https://github.com/curl/curl/files/7475594/output.txt)

**Reply by bagder** [reply]
I agree it appears to be a bug when HTTP/3 is asked for and there's an existing HTTP/2 connection present...

**Reply by bagder** [reply]
The documentation is however clear that libcurl will prefer an existing connection to creating a new one, even if you set a different HTTP version.

**Reply by boytsovea** [reply]
> The documentation is however clear that libcurl will prefer an existing connection to creating a new one, even if you set a different HTTP version.

It seems to me that for switching from HTTP 2 to 3 it is rather unfortunate design choice, especially for backend use cases, because, once established, HTTP/2 connection can stay "alive" in a connection pool for a very long period of time. And more than that, HTTP/2 natively supports multiplexing, which means even less probability to establish new connection with newer version of the protocol. For HTTP 1.1 to 2 transition there were no such problem since both operate on the basis of TCP and in typical scenario second version of the protocol was immediately negotiated during connection handshake.

> I agree it appears to be a bug

I think that in this particular case there is obviously a bug, since request is delivered via HTTP/2 connection over TCP, but declares itself as HTTP/3 request in a starting line. I'm a little surprised, that remote site even recognized that as a valid request.

**Comment by bagder** [comment]
> It seems to me that for switching from HTTP 2 to 3 it is rather unfortunate design choice,

There are existing ways to mitigate and work-around and if they're not enough I would be open for discussing other ways.

>  I'm a little surprised, that remote site even recognized that as a valid request.

You're forgetting that what is displayed in that verbose output is a generated "fake" protocol looking  data. HTTP/2 and HTTP/3 are binary protocols and curl doesn't show the binary protocol exchange there but instead shows the protocol in a HTTP/1-looking style. So it's just *that* code that's wrong, the actual protocol over the wire is done correctly.

**Comment by boytsovea** [comment]
> There are existing ways to mitigate and work-around

Can you please elaborate on this? I'm thinking on limiting connection lifetime via CURLOPT_MAXLIFETIME_CONN, but maybe you can suggest some other options?

**Comment by deepbluev7** [comment]
So I also basically ran into this. I need to either wait until the network died or restart the application. Otherwise it is really hard to know when to start a new session with libcurl (without affecting long lived connections when there is no upgrade available). An easy or multi option like "ACTIVELY_UPGRADE_WITH_RECONNECT" or something would be nice, so that I don't need to handle that myself.

**Reply by bagder** [reply]
1. Most users want a transfer done, and reusing an existing connection is usually the fastest way to that.
2. You can prevent connections from being re-used if you want to enforce a new one to get done
3. You can lower the maximum time curl will allow a connection to stick around in the connection pool until killed

Maybe it would help if you explained your use case.

**Reply by deepbluev7** [reply]
I develop a Matrix client, that does long polling in the background: https://github.com/Nheko-Reborn/nheko/

That means we usually have long running requests (up to 5 minutes usually) and lots of smaller requests at the same time. Uploading or downloading media might also take longer. We would like to reuse the long polling connection for as long as possible to reduce network overhead of just reestablishing the connection. On the other hand switching to http/3 eagerly might significantly improve the experience on mobile.

Currently, because of the long polling, the connection will almost never be reestablished. While we can enforce that by setting a maximum time to reuse a connection, that will mean we unnecessarily reconnect a lot or it takes a very long time to upgrade to http/3.

So ideally we would just reconnect once on http/3, but otherwise keep the connection alive as long as possible. The workarounds are not that horrible. Even requiring a restart to use http/3 is okay. But it just doesn't feel like the right tradeoff. I would rather use a new connection once as soon as we know that http/3 is available.

I hope that makes sense? It is not really something with a high priority in any case!

**Reply by bagder** [reply]
Right, it sounds as if you would want the `CURLALTSVC_IMMEDIATELY` option I've added as a possible future improvement over at https://curl.se/docs/alt-svc.html#TODO

**Reply by deepbluev7** [reply]
Yes, that sounds perfect!
